src/utilities/attachments-manager.ts

```diff
@@ -0,0 +1,105 @@
+  private learnerId?: string;
+  private firebaseJwt?: string;
+  private tokenServiceClient: TokenServiceClient;
+  private resources: Record<string, S3Resource> = {};
+
+  constructor(learnerId: string, firebaseJwt?: string) {
+    this.learnerId = learnerId;
+    this.firebaseJwt = firebaseJwt;
+
+    const env = firebaseAppName() === "report-service-pro" ? "production" : "staging";
+    this.tokenServiceClient = new TokenServiceClient({ env, jwt: firebaseJwt });
+  }
+
+  public isAnonymous() {
+    // The client will be anonymous if firebaseJwt undefined
+    return !this.firebaseJwt;
+  }
+
+  public getSessionId() {
+    return this.sessionId;
+  }
+
+  public async createFolder(interactiveId: string): Promise<IWritableAttachmentsFolder> {
+    const folderResource = await this.tokenServiceClient.createResource({
+      tool: kTokenServiceToolName,
+      type: "s3Folder",
+      name: `${this.learnerId}-${interactiveId}`,
+      description: "attachment",
+      accessRuleType: this.isAnonymous() ? "readWriteToken" : "user"
+    });
+    this.resources[folderResource.id] = folderResource as S3Resource;
+    return {
+      id: folderResource.id,
+      readWriteToken: this.tokenServiceClient.getReadWriteToken(folderResource) || undefined
+    };
+  }
+
+  public async getSignedWriteUrl(
+    folder: IAttachmentsFolder, name: string, expires = kDefaultWriteExpirationSec
+  ): Promise<[string, IReadableAttachmentInfo]> {
+    const folderResource = await this.getFolderResource(folder);
+    const publicPath = this.tokenServiceClient.getPublicS3Path(folderResource, `${this.sessionId}/${name}`);
+    const url = await this.getSignedUrlFromPublicPath(folderResource, publicPath, "putObject", expires);
+    return [url, { folder: { id: folder.id }, publicPath }];
+  }
+
+  public async getSignedReadUrl(attachmentInfo: IReadableAttachmentInfo, expires = kDefaultReadExpirationSec) {
+    const { folder, publicPath } = attachmentInfo;
+    const folderResource = await this.getFolderResource(folder);
+    return this.getSignedUrlFromPublicPath(folderResource, publicPath, "getObject", expires);
+  }
+
+  private async getFolderResource(folder: IAttachmentsFolder): Promise<S3Resource> {
+    let folderResource: S3Resource = this.resources[folder.id];
+    if (!folderResource) {
+      folderResource = await this.tokenServiceClient.getResource(folder.id) as S3Resource;
+      this.resources[folderResource.id] = folderResource;
+    }
+    return folderResource;
+  }
+
+  private getCredentials(folder: IAttachmentsFolder): Promise<Credentials> {
+    const readWriteToken = isWritableAttachmentsFolder(folder) ? folder.readWriteToken : undefined;
+    return this.tokenServiceClient.getCredentials(folder.id, readWriteToken);
+  }
+
+  private async getSignedUrlFromPublicPath(folderResource: S3Resource, publicPath: string, operation: string, expires: number) {
+    const credentials = await this.getCredentials(folderResource);
```

Comment:
This looks like a bug. The passed in `folderResource` is an `S3Resource` from the token-service. But the argument of the getCredentials function is an IAttachmentsFolder. Looking at runtime inspection (at least in the case I inspected) the actual `folderResource` is an `S3Resource`. It doesn't have `folder.readWriteToken`.

Reply:
Good catch -- fixed.